### PR TITLE
✨ Add new API status for the HetznerBareMetalMachineStatus

### LINKED ISSUE #941: Follow phase of Machine in HetznerBareMetalMachine object
## Description

We should have a phase in the status of the HetznerBareMetalMachine object which is optional and omitempty. This phase should be similar to the phases a clusterv1.Machine object goes through and should depend on the host object's provisioningState. 

We should also update the printcolumns accordingly

### PR DESCRIPTION
**Pull request by aniruddha2000**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #941 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

**Comment by janiskemper** [comment]
don't forget the printcolumn :)

**Review by janiskemper — CHANGES_REQUESTED** [review]
I believe that the statuses are not correct. Can you recheck and make sure that if
a) host has provisioning state != provisioned, then we have "provisioning"
b) host has provisioning state == provisioned, then we have "running"?

**Comment by janiskemper** [comment]
and please also add unit tests to the hetznerbaremetal_controller to check that you actually set "running" after the host is provisioned. 
And if there is a test that checks for host == "provisioning", then you can check whether the HetznerBareMetalMachine has phase== provisioning as well

File: pkg/services/baremetal/baremetal/baremetal.go [hunk 1]
```diff
@@ -306,6 +312,12 @@ func (s *Service) associate(ctx context.Context) error {
 	}
 
 	s.ensureMachineAnnotation(host)
+
+	s.scope.BareMetalMachine.Status.Phase = clusterv1.MachinePhaseProvisioning
+	if host.Spec.Status.ProvisioningState == infrav1.StateProvisioned {
+		s.scope.BareMetalMachine.Status.Phase = clusterv1.MachinePhaseRunning
```
Comment (aniruddha2000) [inline pkg/services/baremetal/baremetal/baremetal.go:318]: If we make the phase `running` here do we need to make it again running in the reconcile function above?

Comment (janiskemper) [reply pkg/services/baremetal/baremetal/baremetal.go:318]: the only place where we decide whether a host is ready or not is in setProviderID.. There you can decide whether something is running or not https://github.com/syself/cluster-api-provider-hetzner/blob/ani%2Fissues%2F941/pkg/services/baremetal/baremetal/baremetal.go#L549

File: controllers/hetznerbaremetalmachine_controller_test.go [hunk 2]
```diff
@@ -431,6 +431,22 @@ var _ = Describe("HetznerBareMetalMachineReconciler", func() {
...
 			})
+
+			It("checks the hetznerBareMetalMachine status running phase", func() {
+				By("making sure that machine is in running state")
+				Eventually(func() bool {
+					if err := testEnv.Get(ctx, key, bmMachine); err != nil {
+						return false
+					}
+
+					if err := testEnv.Get(ctx, hostKey, host); err != nil {
```
Comment (janiskemper) [inline controllers/hetznerbaremetalmachine_controller_test.go:442]: you don't need the host here.. You should clean this up

File: controllers/hetznerbaremetalmachine_controller_test.go [hunk 3]
```diff
@@ -490,6 +506,18 @@ var _ = Describe("HetznerBareMetalMachineReconciler", func() {
 				return isPresentAndFalseWithReason(key, bmMachine, infrav1.HostAssociateSucceededCondition, infrav1.NoAvailableHostReason)
 			}, timeout).Should(BeTrue())
 		})
+
+		It("checks the hetznerBareMetalMachine status pending phase", func() {
+			By("making sure that machine is in running state")
```
Comment (janiskemper) [inline controllers/hetznerbaremetalmachine_controller_test.go:511]: what is this supposed to mean?

File: pkg/services/baremetal/baremetal/baremetal.go [hunk 4]
```diff
@@ -116,6 +117,7 @@ func (s *Service) Reconcile(ctx context.Context) (res reconcile.Result, err erro
 	if err := s.setProviderID(ctx); err != nil {
 		return res, fmt.Errorf("failed to set providerID: %w", err)
 	}
+	s.scope.Logger.Info("phase of hetznerBareMetalMachine its here", "phase", s.scope.BareMetalMachine.Status.Phase)
```
Comment (janiskemper) [inline pkg/services/baremetal/baremetal/baremetal.go]: cleanup

File: api/v1beta1/hetznerbaremetalmachine_types.go [hunk 5]
```diff
@@ -265,6 +265,11 @@ type HetznerBareMetalMachineStatus struct {
 	// +optional
 	Ready bool `json:"ready"`
 
+	// Phase represents the current phase of HetznerBareMetalMachineStatus actuation.
+	// E.g. Pending, Running, Terminating, Failed etc.
```
Comment (janiskemper) [inline api/v1beta1/hetznerbaremetalmachine_types.go:269]: Terminating and failed is something we don't even use.. Maybe adjust this comment

**Review by janiskemper — CHANGES_REQUESTED** [review]

**Review by janiskemper — APPROVED** [review]

**Pull request merged**